hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/handlers/TssVoteHandler.java

```diff
@@ -53,5 +62,71 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
+        final var txBody = context.body().tssVoteOrThrow();
+        final var tssBaseStore = context.storeFactory().writableStore(WritableTssStore.class);
+        final var nodeId = context.networkInfo().selfNodeInfo().nodeId();
+        final TssVoteMapKey tssVoteMapKey = new TssVoteMapKey(txBody.targetRosterHash(), nodeId);
+        if (tssBaseStore.exists(tssVoteMapKey)) {
+            // Duplicate vote
+            return;
+        }
+
+        if (!TssVoteHandler.hasReachedThreshold(txBody, context)) {
+            tssBaseStore.put(tssVoteMapKey, txBody);
+        }
+    }
+
+    /**
+     * Check if the threshold number of votes (totaling at least 1/3 of weight) have already been received for the
+     * candidate roster, all with the same vote byte array.
+     *
+     * @param tssVoteTransaction the TssVoteTransaction to check
+     * @param context the HandleContext
+     * @return true if the threshold has been reached, false otherwise
+     */
+    public static boolean hasReachedThreshold(
+            @NonNull final TssVoteTransactionBody tssVoteTransaction, @NonNull final HandleContext context) {
+        final var rosterStore = context.storeFactory().readableStore(ReadableRosterStore.class);
+
+        final Roster activeRoster = rosterStore.getActiveRoster();
+        if (activeRoster == null) {
+            throw new IllegalArgumentException("No active roster found");
+        }
+
+        // Get all votes for the active roster
+        final Map<RosterEntry, TssVoteTransactionBody> voteByNode = new HashMap<>();
+
+        // Get the target roster from the TssVoteTransactionBody
+        final Bytes targetRosterHash = tssVoteTransaction.targetRosterHash();
+
+        // Also get the total active roster weight
+        long activeRosterTotalWeight = 0;
+
+        final var tssBaseStore = context.storeFactory().writableStore(WritableTssStore.class);
+        // For every node in the active roster, check if there is a vote for the target roster hash
+        for (final RosterEntry rosterEntry : rosterStore.getActiveRoster().rosterEntries()) {
+            activeRosterTotalWeight += rosterEntry.weight();
+            final TssVoteMapKey tssVoteMapKey = new TssVoteMapKey(targetRosterHash, rosterEntry.nodeId());
+            if (tssBaseStore.exists(tssVoteMapKey)) {
+                voteByNode.put(rosterEntry, tssBaseStore.getVote(tssVoteMapKey));
+            }
+        }
+
+        // Initialize a counter for the total weight of votes with the same vote byte array
+        long voteWeight = 0L;
+
+        // Iterate over the votes which has the same target roster hash
+        for (final RosterEntry rosterEntryKey : voteByNode.keySet()) {
+            final TssVoteTransactionBody vote = voteByNode.get(rosterEntryKey);
+            // If the vote byte array matches the one in the TssVoteTransaction, add the weight of the vote to the
+            // counter
+            if (vote.tssVote().equals(tssVoteTransaction.tssVote())) {
```

Comment:
Given this is executing on the transaction handling thread and every micro second counts, couldn't we skip this second for loop and just do this inside the body of the if statement on line 110?   We wouldn't need to allocate the memory of creating a voteByNode map or the cost of insert and lookup into it.

Reply:
Agreed. Fixed it. Thanks @edward-swirldslabs